rust/src/http2/http2.rs

```diff
@@ -1002,19 +1041,58 @@ impl HTTP2State {
                                         dinput,
+                                        &mut output,
                                         dir,
                                         sfcm,
                                         over,
-                                        flow).is_err() {
-                                        self.set_event(HTTP2Event::FailedDecompression);
+                                        flow) {
+                                    Ok(_) => {
+                                        if !tx_same.doh.is_empty() {
+                                            if over {
+                                                // fix DNS/TCP header length field
+                                                tx_same.doh[0] = ((tx_same.doh.len() - 2) >> 8) as u8;
+                                                tx_same.doh[1] = ((tx_same.doh.len() - 2) & 0xFF) as u8;
+                                                self.layered.push(tx_same.doh.to_vec());
+                                                if self.dns_state.is_none() {
+                                                    unsafe {
+                                                        self.dns_state = Some(rs_dns_state_new(std::ptr::null_mut(), ALPROTO_HTTP2));
+                                                    }
+                                                }
+                                                unsafe {
+                                                    AppLayerParserStateSetFlag(
+                                                        pstate,
+                                                        APP_LAYER_PARSER_LAYERED_PACKET,
+                                                    );
+                                                }
+                                            }
+                                        }
+                                    }
+                                        _ => {
+                                            self.set_event(HTTP2Event::FailedDecompression);
+                                        }
                                     }
                                 }
                             }
                             None => panic!("no SURICATA_HTTP2_FILE_CONFIG"),
                         }
                     }
+                    if let Some(doh) = odoh {
+                        self.layered.push(doh);
+                        unsafe {
+                            AppLayerParserStateSetFlag(
+                                pstate,
+                                APP_LAYER_PARSER_LAYERED_PACKET,
+                            );
+                        }
+                        if self.dns_state.is_none() {
+                            unsafe {
+                                self.dns_state = Some(rs_dns_state_new(std::ptr::null_mut(), ALPROTO_HTTP2));
+                            }
+                        }
```

Comment:
⚠️  This is a key first step :
we
- set a flag in pstate  (so that AppLayerParserParse knows that it should deque packets)
- push the to-be pseudo-packet payload in HTTP2 state
- create a DNS state if there is not one yet